### ISSUE TITLE
SOAP fails on large runs due to halo catalogue arrays too large to MPI_GATHER()

### ISSUE THREAD

**Issue by jchelly** [SWIFTSIM/SOAP#46]
Currently SOAP reads the input halo catalogue in parallel across all MPI ranks, but then gathers it on one rank to split it up into chunks. This fails on the L2800N10080 FLAMINGO run because it involves an allgather() where the count doesn't fit in an int.

The quick solution would be to implement our own gather which can handle large arrays using sends/recvs.

A more scalable solution would be to keep the halo catalogue distributed over all MPI ranks. When a chunk task is executed it would first have to request its assigned halos from whichever MPI rank they're stored on. Each rank would need to start a thread to respond to these requests. Since the chunk tasks are long running this thread could sleep most of the time and just check for messages periodically. The thread that hands out chunk tasks to nodes already works like this.
